File: backend/libexecution/libint.ml
open Core_kernel
open Lib
open Types.RuntimeT
module RT = Runtime

(* type coerces one list to another using a function *)
let list_coerce ~(f : dval -> 'a option) (l : dval list) :
    ('a list, dval list * dval) Result.t =
  l
  |> List.map ~f:(fun dv ->
         match f dv with Some v -> Result.Ok v | None -> Result.Error (l, dv))
  |> Result.all


let ( >>| ) = Result.( >>| )

let fns : fn list =
  [ { prefix_names = ["Int::mod"]
    ; infix_names = ["%"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TInt
    ; description =
        "Return `a` % `b`, the modulus of a and b. This is the integer remainder left when `a` is divided by `b`. For example, `15 % 6 = 3`."
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] -> DInt (Dint.( % ) a b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::add"]
    ; infix_names = ["+"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TInt
    ; description = "Adds two integers together"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DInt (Dint.( + ) a b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::add to add Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::add to add Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [(DStr _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a String, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use ++ or String::append to join Strings or parse Strings to Ints with String::toInt."
                )
          | state, [_; (DStr _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a String, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use ++ or String::append to join Strings or parse Strings to Ints with String::toInt."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::subtract"]
    ; infix_names = ["-"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TInt
    ; description = "Subtracts two integers"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DInt (Dint.( - ) a b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::subtract to subtract Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::subtract to subtract Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::multiply"]
    ; infix_names = ["*"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TInt
    ; description = "Multiplies two integers"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DInt (Dint.( * ) a b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::multiply to multiply Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::multiply to multiply Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::power"]
    ; infix_names = ["^"]
    ; parameters = [par "base" TInt; par "exponent" TInt]
    ; return_type = TInt
    ; description = "Raise `base` to the power of `exponent`"
    ; func =
        InProcess
          (function
          | _, [DInt base; DInt exp] ->
              DInt (Dint.pow base exp)
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::divide"]
    ; infix_names = []
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TInt
    ; description = "Divides two integers"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DInt (Dint.( / ) a b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::divide to divide Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::divide to divide Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::greaterThan"]
    ; infix_names = [">"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns true if a is greater than b"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DBool (a > b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::greaterThan to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::greaterThan to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::greaterThanOrEqualTo"]
    ; infix_names = [">="]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns true if a is greater than or equal to b"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DBool (a >= b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::greaterThanOrEqualTo to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::greaterThanOrEqualTo to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::lessThan"]
    ; infix_names = ["<"]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns true if a is less than b"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DBool (a < b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::lessThan to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::lessThan to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::lessThanOrEqualTo"]
    ; infix_names = ["<="]
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns true if a is less than or equal to b"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              DBool (a <= b)
          | state, [(DFloat _ as a); _] ->
              DError
                ( SourceNone
                , "The first param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::lessThanOrEqualTo to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | state, [_; (DFloat _ as b)] ->
              DError
                ( SourceNone
                , "The second param ("
                  ^ Dval.to_developer_repr_v0 b
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::lessThanOrEqualTo to compare Floats or use Float::floor to truncate Floats to Ints."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::random"]
    ; infix_names = []
    ; parameters = [par "start" TInt; par "end" TInt]
    ; return_type = TInt
    ; description = "Returns a random integer between a and b (inclusive)"
    ; func =
        InProcess
          (function
          (*( +1 as Random.int is exclusive *)
          | _, [DInt a; DInt b] ->
              let open Dint in
              DInt (a + one + Dint.random (b - a))
          | args ->
              fail args)
    ; preview_safety = Unsafe
    ; deprecated = true }
  ; { prefix_names = ["Int::random_v1"]
    ; infix_names = []
    ; parameters = [par "start" TInt; par "end" TInt]
    ; return_type = TInt
    ; description =
        "Returns a random integer between `start` and `end` (inclusive)."
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] ->
              let open Dint in
              (* upper+1 because as Random.int is exclusive *)
              let lower, upper = if a > b then (b, a + one) else (a, b + one) in
              DInt (lower + Dint.random (upper - lower))
          | args ->
              fail args)
    ; preview_safety = Unsafe
    ; deprecated = false }
  ; { prefix_names = ["Int::sqrt"]
    ; infix_names = []
    ; parameters = [par "a" TInt]
    ; return_type = TFloat
    ; description = "Get the square root of an Int"
    ; func =
        InProcess
          (function
          | _, [DInt a] ->
              DFloat (Dint.to_float a |> sqrt)
          | state, [(DFloat _ as a)] ->
              DError
                ( SourceNone
                , "The param ("
                  ^ Dval.to_developer_repr_v0 a
                  ^ ") is a Float, but "
                  ^ state.executing_fnname
                  ^ " only works on Ints. Use Float::sqrt to take the square root of Floats or use Float::floor to truncate the Float to an Int."
                )
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::toFloat"]
    ; infix_names = []
    ; parameters = [par "a" TInt]
    ; return_type = TFloat
    ; description = "Converts an Int to a Float"
    ; func =
        InProcess
          (function
          | _, [DInt a] -> DFloat (Dint.to_float a) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::sum"]
    ; infix_names = []
    ; parameters = [par "a" TList]
    ; return_type = TInt
    ; description = "Returns the sum of all the ints in the list"
    ; func =
        InProcess
          (function
          | _, [DList l] ->
              l
              |> list_coerce ~f:Dval.to_int
              >>| List.fold_left ~f:Dint.( + ) ~init:Dint.zero
              >>| (fun x -> DInt x)
              |> Result.map_error ~f:(fun (result, example_value) ->
                     RT.error
                       ~actual:(DList result)
                       ~result:(DList result)
                       ~long:
                         ( "Int::sum requires all values to be integers, but "
                         ^ Dval.to_developer_repr_v0 example_value
                         ^ " is a "
                         ^ Dval.pretty_tipename example_value )
                       ~expected:"every list item to be an int "
                       "Sum expects you to pass a list of ints")
              |> Result.ok_exn
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::max"]
    ; infix_names = []
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns the higher of a and b"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] -> DInt (Dint.max a b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Int::min"]
    ; infix_names = []
    ; parameters = [par "a" TInt; par "b" TInt]
    ; return_type = TBool
    ; description = "Returns the lower of `a` and `b`"
    ; func =
        InProcess
          (function
          | _, [DInt a; DInt b] -> DInt (Dint.min a b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false } ]

File: backend/src/LibExecutionStdLib/LibInt.fs
module LibExecutionStdLib.LibInt

open System.Threading.Tasks
open System.Numerics
open FSharp.Control.Tasks

open LibExecution.RuntimeTypes
open Prelude

module Errors = LibExecution.Errors

let fn = FQFnName.stdlibFnName

let err (str : string) = Ply(Dval.errStr str)

let incorrectArgs = Errors.incorrectArgs


let fns : List<BuiltInFn> =
  [ { name = fn "Int" "mod" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description =
        "Returns the result of wrapping <param a> around so that {{0 <= res < b}}.

         The modulus <param b> must be 0 or negative.

         Use <fn Int::remainder> if you want the remainder after division, which has
         a different behavior for negative numbers."
      fn =
        (function
        | _, [ DInt v; DInt m as mdv ] ->
          if m <= 0L then
            err (Errors.argumentWasnt "positive" "b" mdv)
          else
            // dotnet returns negative mods, but OCaml did positive ones
            let result = v % m
            let result = if result < 0L then m + result else result
            Ply(DInt(result))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "%"
      previewable = Pure
      // TODO: Deprecate this when we can version infix operators and when infix operators support Result return types (https://github.com/darklang/dark/issues/4267)
      // The current function returns DError (it used to rollbar) on negative `b`.
      deprecated = NotDeprecated }


    // See above for when to uncomment this
    // TODO: A future version should support all non-zero modulus values and should include the infix "%"
    // { name = fn "Int" "mod" 1
    //   parameters = [ Param.make "value" TInt ""; Param.make "modulus" TInt "" ]
    //   returnType = TResult(TInt, TStr)
    //   description =
    //     "Returns the result of wrapping <param value> around so that {{0 <= res < modulus}}, as a <type Result>.
    //      If <param modulus> is positive, returns {{Ok res}}. Returns an {{Error}} if <param modulus> is {{0}} or negative.
    //     Use <fn Int::remainder> if you want the remainder after division, which has a different behavior for negative numbers."
    //   fn =
    //     (function
    //     | _, [ DInt v; DInt m ] ->
    //       (try
    //         Ply(DResult(Ok(DInt(v % m))))
    //        with
    //        | e ->
    //          if m <= 0L then
    //            Ply(
    //              DResult(
    //                Error(
    //                  DStr(
    //                    "`modulus` must be positive but was "
    //                    + LibExecution.DvalReprDeveloper.toRepr (DInt m)
    //                  )
    //                )
    //              )
    //            )
    //          else // In case there's another failure mode, rollbar
    //            Exception.raiseInternal "Unexpected failiure mode" [] e)
    //     | _ -> incorrectArgs ())
    //   sqlSpec = NotYetImplemented
    //   previewable = Pure
    //   deprecated = NotDeprecated }


    { name = fn "Int" "remainder" 0
      parameters = [ Param.make "value" TInt ""; Param.make "divisor" TInt "" ]
      returnType = TResult(TInt, TStr)
      description =
        "Returns the integer remainder left over after dividing <param value> by
         <param divisor>, as a <type Result>.

         For example, {{Int::remainder 15 6 == Ok 3}}. The remainder will be
         negative only if {{<var value> < 0}}.

         The sign of <param divisor> doesn't influence the outcome.

         Returns an {{Error}} if <param divisor> is {{0}}."
      fn =
        (function
        | _, [ DInt v; DInt d ] ->
          (try
            v % d |> DInt |> Ok |> DResult |> Ply
           with
           | e ->
             if d = 0L then
               Ply(DResult(Error(DStr($"`divisor` must be non-zero"))))
             else
               Exception.raiseInternal
                 "unexpected failure case in Int::remainder"
                 [ "v", v; "d", d ]
                 e)
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplemented
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "add" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Adds two integers together"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DInt(a + b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "+"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "subtract" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Subtracts two integers"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DInt(a - b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "-"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "multiply" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Multiplies two integers"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DInt(a * b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "*"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "power" 0
      parameters = [ Param.make "base" TInt ""; Param.make "exponent" TInt "" ]
      returnType = TInt
      description = "Raise <param base> to the power of <param exponent>"
      fn =
        (function
        | _, [ DInt number; DInt exp as expdv ] ->
          (try
            if exp < 0L then
              err (Errors.argumentWasnt "positive" "exponent" expdv)
            // Handle some edge cases around 1. We want to make this match
            // OCaml, so we have to support an exponent above int32, but
            // below int63. This only matters for 1 or -1, and otherwise a
            // number raised to an int63 exponent wouldn't fit in an int63
            else if number = 1L then
              Ply(DInt(1L))
            else if number = -1L && exp % 2L = 0L then
              Ply(DInt(1L))
            else if number = -1L then
              Ply(DInt(-1L))
            else
              (bigint number) ** (int exp) |> int64 |> DInt |> Ply
           with
           | _ -> err "Error raising to exponent")
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "^"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "divide" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Divides two integers"
      fn =
        (function
        | _, [ DInt a; DInt b ] ->
          if b = 0L then err "Division by zero" else Ply(DInt(a / b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "/"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "absoluteValue" 0
      parameters = [ Param.make "a" TInt "" ]
      returnType = TInt
      description =
        "Returns the absolute value of <param a> (turning negative inputs into positive outputs)"
      fn =
        (function
        | _, [ DInt a ] -> Ply(DInt(abs a))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "negate" 0
      parameters = [ Param.make "a" TInt "" ]
      returnType = TInt
      description = "Returns the negation of <param a>, {{-a}}"
      fn =
        (function
        | _, [ DInt a ] -> Ply(DInt(-a))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "greaterThan" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TBool
      description = "Returns {{true}} if <param a> is greater than <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DBool(a > b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp ">"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "greaterThanOrEqualTo" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TBool
      description =
        "Returns {{true}} if <param a> is greater than or equal to <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DBool(a >= b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp ">="
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "lessThan" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TBool
      description = "Returns {{true}} if <param a> is less than <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DBool(a < b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "<"
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "lessThanOrEqualTo" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TBool
      description =
        "Returns {{true}} if <param a> is less than or equal to <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DBool(a <= b))
        | _ -> incorrectArgs ())
      sqlSpec = SqlBinOp "<="
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "random" 1
      parameters = [ Param.make "start" TInt ""; Param.make "end" TInt "" ]
      returnType = TInt
      description =
        "Returns a random integer between <param start> and <param end> (inclusive)"
      fn =
        (function
        | _, [ DInt a; DInt b ] ->
          let lower, upper = if a > b then (b, a) else (a, b)
          // The next line is a fix to correct work when an upper bound equal '1'.
          // It is need because System.Random.NexInt64(Int64) _always_ returns 0 for 1
          // line from the official doc: "The _exclusive_ upper bound of the random number to be generated"
          let correction : int64 = if upper = 1 then 2 else 0
          lower + randomSeeded().NextInt64(upper - lower + correction) |> DInt |> Ply
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = NotDeprecated }


    { name = fn "Int" "sqrt" 0
      parameters = [ Param.make "a" TInt "" ]
      returnType = TFloat
      description = "Get the square root of an <type Int>"
      fn =
        (function
        | _, [ DInt a ] -> Ply(DFloat(sqrt (float a)))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "toFloat" 0
      parameters = [ Param.make "a" TInt "" ]
      returnType = TFloat
      description = "Converts an <type Int> to a <type Float>"
      fn =
        (function
        | _, [ DInt a ] -> Ply(DFloat(float a))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "sum" 0
      parameters = [ Param.make "a" (TList TInt) "" ]
      returnType = TInt
      description = "Returns the sum of all the ints in the list"
      fn =
        (function
        | _, [ DList l as ldv ] ->
          let ints =
            List.map
              (fun i ->
                match i with
                | DInt it -> it
                | _ ->
                  Exception.raiseCode (Errors.argumentWasnt "a list of ints" "a" ldv))
              l

          let sum = List.fold (fun acc elem -> acc + elem) 0L ints
          Ply(DInt sum)
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "max" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Returns the higher of <param a> and <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DInt(max a b))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "min" 0
      parameters = [ Param.make "a" TInt ""; Param.make "b" TInt "" ]
      returnType = TInt
      description = "Returns the lower of <param a> and <param b>"
      fn =
        (function
        | _, [ DInt a; DInt b ] -> Ply(DInt(min a b))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "clamp" 0
      parameters =
        [ Param.make "value" TInt ""
          Param.make "limitA" TInt ""
          Param.make "limitB" TInt "" ]
      returnType = TInt
      description =
        "If <param value> is within the range given by <param limitA> and <param
         limitB>, returns <param value>.

         If <param value> is outside the range, returns <param limitA> or <param
         limitB>, whichever is closer to <param value>.

         <param limitA> and <param limitB> can be provided in any order."
      fn =
        (function
        | _, [ DInt v; DInt a; DInt b ] ->
          let min, max = if a < b then (a, b) else (b, a)

          if v < min then Ply(DInt min)
          else if v > max then Ply(DInt max)
          else Ply(DInt v)
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "parse" 0
      parameters = [ Param.make "s" TStr "" ]
      returnType = TResult(TInt, TStr)
      description = "Returns the <type int> value of a <type string>"
      fn =
        (function
        | _, [ DStr s ] ->
          (try
            s |> System.Convert.ToInt64 |> DInt |> Ok |> DResult |> Ply
           with
           | _e ->
             $"Expected to parse string with only numbers, instead got \"{s}\""
             |> DStr
             |> Error
             |> DResult
             |> Ply)
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplemented
      previewable = Pure
      deprecated = NotDeprecated }


    { name = fn "Int" "toString" 0
      parameters = [ Param.make "int" TInt "" ]
      returnType = TStr
      description = "Stringify <param int>"
      fn =
        (function
        | _, [ DInt int ] -> Ply(DStr(string int))
        | _ -> incorrectArgs ())
      sqlSpec = NotQueryable
      previewable = Pure
      deprecated = NotDeprecated }


    ]
